{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8f9b9f3a-6698-453c-a3be-40c567409e9c",
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "source": [
    "# PairOT Tutorial"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f9a2abb4-6bd0-4b10-94fa-89444faf010b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: XLA_PYTHON_CLIENT_MEM_FRACTION=.99\n"
     ]
    }
   ],
   "source": [
    "# Allow JAX to use all available GPU memory\n",
    "%env XLA_PYTHON_CLIENT_MEM_FRACTION=.99"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6bebb7a2-e9f9-4664-9e92-ed34ac3aae45",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "import anndata as ad\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pairot as pr\n",
    "import scanpy as sc\n",
    "\n",
    "from IPython.display import display, Image"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d13b75a3-6e04-433e-8ada-c10d92b570b2",
   "metadata": {},
   "source": [
    "## 0. Required software and data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dbcb651e-7503-4f7a-8b82-7a532300bf7d",
   "metadata": {},
   "source": [
    "**Docker containers:**\n",
    "  * pairOT Docker container: https://hub.docker.com/repository/docker/felix0097/pairot/general\n",
    "  * `docker pull felix0097/pairot:full_v1`\n",
    "\n",
    "**Datasets:**\n",
    "  * Van der Wijst (Query dataset): https://cellxgene.cziscience.com/collections/7d7cabfd-1d1f-40af-96b7-26a0825a306d\n",
    "  * Asian Immune Diversity Atlas (Reference dataset): https://celltype.info/project/336/dataset/591/label/71663\n",
    "\n",
    "**Required Hardware:**\n",
    "  * We recommend a GPU with at least 40GB of VRAM\n",
    "  * We used a Nvidia A40 GPU for this tutorial (48GB VRAM)\n",
    "  * Moreover, for the `Data preparation` section, we recommend around 200GB of system memory"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87fbb0b7-3b8a-46b1-aa64-f7fc44d2311f",
   "metadata": {},
   "source": [
    "## 1. Data preparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8b6f6e34-ae02-43d3-bf88-d17ca2899938",
   "metadata": {},
   "outputs": [],
   "source": [
    "query_dataset = \"7d7cabfd-1d1f-40af-96b7-26a0825a306d\"\n",
    "ref_dataset = \"ced320a1-29f3-47c1-a735-513c7084d508_CAP\"\n",
    "\n",
    "raw_data_path = Path(\"/vol/data/dataset-similarity/preprocessed\")\n",
    "preproc_data_path = Path(\"/vol/data/dataset-similarity/cache\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "caa22a45-61db-4514-adc1-22e8d45da941",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load h5ad files\n",
    "adata_query = sc.read_h5ad(raw_data_path / f\"{query_dataset}.h5ad\")\n",
    "adata_ref = sc.read_h5ad(raw_data_path / f\"{ref_dataset}.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c133918-24f0-4916-91f0-a2c0f660a340",
   "metadata": {},
   "source": [
    "**Gene space alignment:**\n",
    "* SConnect only takes genes into account that are expressed in both the query and reference dataset. Hence, it is crucial that gene names are uniform across both datasets.\n",
    "* Genes should be named in accordance with the gene names listed on [genenames.org](https://genenames.org/tools/search/#!/?query=&filter=document_type:gene)\n",
    "\n",
    "\n",
    "**Filtering of differentially expressed genes:**\n",
    "* We filter out uninformative genes from the differentially expression testing results.\n",
    "* Hence, it is crucial that the gene names used are in accordance with the gene names listed on [genenames.org](https://genenames.org/tools/search/#!/?query=&filter=document_type:gene)\n",
    "* We use two filters:\n",
    "  * We subset all genes to the genes listed in `OFFICIAL_GENES[\"feature_name\"]`\n",
    "  * We filter out uninformative genes listed in `FILTERED_GENES[\"feature_name\"]`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c452170e-b0a2-424a-8538-a5f5a4060b8c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>origin</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature_name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RNU6-992P</th>\n",
       "      <td>Approved symbol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C9orf71</th>\n",
       "      <td>Previous symbols</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MLIV</th>\n",
       "      <td>Alias symbols</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNX26</th>\n",
       "      <td>Previous symbols</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-mir-653</th>\n",
       "      <td>Alias symbols</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR12D1P</th>\n",
       "      <td>Previous symbols</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DEFB130B</th>\n",
       "      <td>Approved symbol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC00269</th>\n",
       "      <td>Approved symbol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA5SP167</th>\n",
       "      <td>Approved symbol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HUSI</th>\n",
       "      <td>Alias symbols</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        origin\n",
       "feature_name                  \n",
       "RNU6-992P      Approved symbol\n",
       "C9orf71       Previous symbols\n",
       "MLIV             Alias symbols\n",
       "SNX26         Previous symbols\n",
       "hsa-mir-653      Alias symbols\n",
       "OR12D1P       Previous symbols\n",
       "DEFB130B       Approved symbol\n",
       "LINC00269      Approved symbol\n",
       "RNA5SP167      Approved symbol\n",
       "HUSI             Alias symbols"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# All DEGs (differentially expressed genes) are subset to the genes list in OFFICAL_GENES\n",
    "# Genes that are not list there are discarded\n",
    "pr.pp.OFFICIAL_GENES.set_index(\"feature_name\").sample(10, random_state=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5e7c8ca0-36ff-4db9-ae3c-fad232aed2c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>origin</th>\n",
       "      <th>filter_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KCNMA1-AS2</th>\n",
       "      <td>Approved symbol</td>\n",
       "      <td>IncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLRK1-AS1</th>\n",
       "      <td>Approved symbol</td>\n",
       "      <td>IncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>U3-55K</th>\n",
       "      <td>Alias symbols</td>\n",
       "      <td>ribosomal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-616M22.2</th>\n",
       "      <td>Alias symbols</td>\n",
       "      <td>IncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC00900</th>\n",
       "      <td>Approved symbol</td>\n",
       "      <td>IncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MTATP6P27</th>\n",
       "      <td>Approved symbol</td>\n",
       "      <td>mitochondrial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RPL5P30</th>\n",
       "      <td>Approved symbol</td>\n",
       "      <td>ribosomal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C20orf61</th>\n",
       "      <td>Previous symbols</td>\n",
       "      <td>IncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RPSAP46</th>\n",
       "      <td>Approved symbol</td>\n",
       "      <td>ribosomal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PPP4R3B-DT</th>\n",
       "      <td>Approved symbol</td>\n",
       "      <td>IncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         origin    filter_type\n",
       "feature_name                                  \n",
       "KCNMA1-AS2      Approved symbol         IncRNA\n",
       "KLRK1-AS1       Approved symbol         IncRNA\n",
       "U3-55K            Alias symbols      ribosomal\n",
       "RP11-616M22.2     Alias symbols         IncRNA\n",
       "LINC00900       Approved symbol         IncRNA\n",
       "MTATP6P27       Approved symbol  mitochondrial\n",
       "RPL5P30         Approved symbol      ribosomal\n",
       "C20orf61       Previous symbols         IncRNA\n",
       "RPSAP46         Approved symbol      ribosomal\n",
       "PPP4R3B-DT      Approved symbol         IncRNA"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# All DEGs (differentially expressed genes) that are listed in FILTERED_GENES are discarded\n",
    "# from differential expression testing results\n",
    "pr.pp.FILTERED_GENES.set_index(\"feature_name\").sample(10, random_state=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "878849fe-f05c-45e5-bd1a-f2c47d05b0d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature_name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TSPAN6</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNMD</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DPM1</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCYL3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf112</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-484N11.1</th>\n",
       "      <td>ENSG00000288321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATXN8</th>\n",
       "      <td>ENSG00000288330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP4-733M16.8</th>\n",
       "      <td>ENSG00000288398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-260M19.4</th>\n",
       "      <td>ENSG00000288459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SMIM42</th>\n",
       "      <td>ENSG00000288460</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>37976 rows √ó 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    feature_id\n",
       "feature_name                  \n",
       "TSPAN6         ENSG00000000003\n",
       "TNMD           ENSG00000000005\n",
       "DPM1           ENSG00000000419\n",
       "SCYL3          ENSG00000000457\n",
       "C1orf112       ENSG00000000460\n",
       "...                        ...\n",
       "RP11-484N11.1  ENSG00000288321\n",
       "ATXN8          ENSG00000288330\n",
       "RP4-733M16.8   ENSG00000288398\n",
       "RP11-260M19.4  ENSG00000288459\n",
       "SMIM42         ENSG00000288460\n",
       "\n",
       "[37976 rows x 1 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# SConnect uses the var.index column\n",
    "# Here, we use gene names and not ENSEMBLE IDs -> this looks good\n",
    "adata_query.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b5243da5-3f75-4453-be0e-0bcc0076f080",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature_name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TSPAN6</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNMD</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DPM1</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCYL3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf112</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP4-669P10.21</th>\n",
       "      <td>ENSG00000288719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-852E15.3</th>\n",
       "      <td>ENSG00000288720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F8A1</th>\n",
       "      <td>ENSG00000288722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-553N16.6</th>\n",
       "      <td>ENSG00000288723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP13-546I2.2</th>\n",
       "      <td>ENSG00000288724</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38114 rows √ó 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    feature_id\n",
       "feature_name                  \n",
       "TSPAN6         ENSG00000000003\n",
       "TNMD           ENSG00000000005\n",
       "DPM1           ENSG00000000419\n",
       "SCYL3          ENSG00000000457\n",
       "C1orf112       ENSG00000000460\n",
       "...                        ...\n",
       "RP4-669P10.21  ENSG00000288719\n",
       "RP11-852E15.3  ENSG00000288720\n",
       "F8A1           ENSG00000288722\n",
       "RP11-553N16.6  ENSG00000288723\n",
       "RP13-546I2.2   ENSG00000288724\n",
       "\n",
       "[38114 rows x 1 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# SConnect uses the var.index column\n",
    "# Here, we use gene names and not ENSEMBLE IDs -> this looks good\n",
    "adata_ref.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "02179e17-6e2f-416f-8e96-60058ed657e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of shared genes: 18172\n",
      "Fraction of genes in official genes list: 0.887\n"
     ]
    }
   ],
   "source": [
    "# Let's run some checks before continuing\n",
    "official_genes = set(pr.pp.OFFICIAL_GENES[\"feature_name\"])\n",
    "\n",
    "\n",
    "# 1. check that there is a decent overlap in the gene spaces of the two dataset\n",
    "expressed_genes_query = set(adata_query.var.index[np.array(adata_query.X.sum(axis=0)).flatten() > 0.0].tolist())\n",
    "expressed_genes_ref = set(adata_ref.var.index[np.array(adata_ref.X.sum(axis=0)).flatten() > 0.0].tolist())\n",
    "intersection = expressed_genes_query & expressed_genes_ref\n",
    "print(f\"Number of shared genes: {len(intersection)}\")\n",
    "\n",
    "\n",
    "# 2. Fraction of genes names that are in the OFFICIAL_GENES list\n",
    "print(f\"Fraction of genes in official genes list: {len(intersection & official_genes) / len(intersection):.3f}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9486c7cc-ae5a-4b56-abb7-9c46df2fd247",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Preprocess datasets, this includes:\n",
    "# 1. Aligning/sort gene space\n",
    "# 2. Do differential expression testing\n",
    "# 3. Select highly variable genes\n",
    "adata_query, adata_ref = pr.pp.preprocess_adatas(\n",
    "    adata_query,\n",
    "    adata_ref,\n",
    "    n_top_genes=750,\n",
    "    cell_type_column_adata1=\"cell_type_author\",\n",
    "    cell_type_column_adata2=\"cell_type_author\",\n",
    "    sample_column_adata1=\"sample_id\",\n",
    "    sample_column_adata2=\"sample_id\",\n",
    ")\n",
    "# Cache preprocessed data\n",
    "adata_query.write_h5ad(preproc_data_path / f\"{query_dataset}.h5ad\")\n",
    "adata_ref.write_h5ad(preproc_data_path / f\"{ref_dataset}.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e10fd92-31e7-4907-9cb9-6890c3b51dc0",
   "metadata": {},
   "source": [
    "**Explanation of method parameters:**\n",
    "* `n_top_genes`: Number of highly variable genes for calculating Spearman correlations\n",
    "  * Based on the analysis of [Crow et al.](https://www.nature.com/articles/s41467-018-03282-0), values between 250 and 750 work well. Choosing a higher value leads to increased computational cost (linear increase in the number of genes) while keeping the method output mostly the same.\n",
    "  * Recommended default value: 750\n",
    "  * This parameter can be used to speed up computation or to save GPU memory.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea1f2968-1970-47a2-8fc2-74b0bb6f8c66",
   "metadata": {},
   "source": [
    "## 2. Initialize SConnect model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dee058e1-ee70-417e-8ee7-d10358c2ebc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_query = ad.read_h5ad(preproc_data_path / f\"{query_dataset}.h5ad\")\n",
    "adata_ref = ad.read_h5ad(preproc_data_path / f\"{ref_dataset}.h5ad\")\n",
    "# Make sure that the genes have the same order in both dataset\n",
    "assert adata_query.var.index.equals(adata_ref.var.index)\n",
    "\n",
    "dataset_map = pr.tl.DatasetMap(adata_query, adata_ref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8c5fabc0-4169-43c0-a87c-37323b27a01b",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_map.init_geom(\n",
    "    batch_size=2048,\n",
    "    epsilon=0.05,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a559d15-25cf-42f6-9581-dcb8ae742d4c",
   "metadata": {},
   "source": [
    "**Explanation of method parameters:**\n",
    "* `batch_size`\n",
    "  * This purely technical parameter affects the GPU memory usage of SConnect and does NOT affect its output. Given GPU memory limitations, it should be chosen as large as possible to fully utilize GPU compute capacity.\n",
    "  * If one receives GPU memory errors later in the code, one should reduce the batch_size parameter.\n",
    "  * (Note: This parameter has nothing to do with single-cell sequencing batches or the ‚Äúbatch size‚Äù used in stochastic gradient descent.)\n",
    "* `epsilon`: regularisation strength of the optimal transport problem\n",
    "  * This parameter adjusts the regularisation strength of the optimal transport problem.\n",
    "  * Intuitively, this parameter affects the uncertainty estimates provided by SConnect. The higher Œµ is, the more conservative the uncertainty parameters become (fewer false positive matches).\n",
    "  * Recommended default value: 0.05\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "e715f371-3870-4100-8412-30f6fef5e90f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "x size: 1.805190716 GB\n",
      "y size: 3.801934496 GB\n"
     ]
    }
   ],
   "source": [
    "# We can print the memory usage of the dataset two count matrices of the query and reference dataset\n",
    "# Those should comfortably fit into GPU memory\n",
    "# If this is not the case, you should reduce the number of highly variable genes\n",
    "# Reducing the number of highly variable genes will also speed up computation\n",
    "# Any value between 250 and 750 should work fairly well\n",
    "\n",
    "print(f\"x size: {dataset_map.geom.x.shape[0] * dataset_map.geom.x.shape[1] * 4 / 1000**3} GB\")\n",
    "print(f\"y size: {dataset_map.geom.y.shape[0] * dataset_map.geom.y.shape[1] * 4 / 1000**3} GB\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a68013a-a7d7-44dc-b59a-0eef2f38ac1e",
   "metadata": {},
   "source": [
    "## 3. Model description & initial sanity checks of the input data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8dddb958-2cd0-41ba-8781-1fcdcbfccbe0",
   "metadata": {},
   "source": [
    "**Model description:**\n",
    "\n",
    "SConnect aims to ‚Äúalign‚Äù (or to ‚Äúconnect‚Äù) cell-type labels between two datasets from separate studies, suggesting similar clusters or potential matches against the reference dataset for each cell type in the query dataset solely based on the underlying transcriptomic signatures. To achieve this objective, we model each dataset as a point cloud, meaning each data point or cell is associated with a gene expression vector (ùòÖ) and a cluster / cell-type label (y). Notably, the clustering or grouping information of individual cells uniformly annotated, not the associated string labels, is provided by the cell-type label y. At its core, SConnect uses optimal transport to connect the point cloud distribution of the query dataset with the point cloud from the reference dataset; thus, the method considers the global structure of the data, compared to just finding the closest neighbors in the reference dataset.\n",
    "\n",
    "The distance between a cell in the query dataset (described by the multi-dimensional gene expression vector $ùòÖ_1$ and cluster label $y_1$) and a cell in the reference dataset (described by the multi-dimensional gene expression vector $ùòÖ_2$ and cluster label $y_2$) is split into two parts. The distance measure (also known as the ‚Äútransport cost‚Äù) is the sum between the distance in gene expression space and the distance in label space:\n",
    "$$\n",
    "distance((ùòÖ_1, y_1), (ùòÖ_2, y_2)) = \\lambda_{feature} \\cdot distance_{gene\\,expression} (ùòÖ_1, ùòÖ_2) + \\lambda_{label} \\cdot distance_{label}(y_1, y_2)\n",
    "$$"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef4216eb-64c3-4708-8425-9e510fcb9878",
   "metadata": {},
   "source": [
    "**Doing an initial sanity check:**\n",
    "\n",
    "First, looking at the above distance measure, we can do an initial sanity check:\n",
    "\n",
    "The optimal transport model relies on a predictive distance measure to accurately align cell types between clusters. Hence, as a quick sanity check we can plot the label distance matrix $distance_{label}(y_1, y_2)$ between cell type clusters. The label distance is computed by looking at the similarity of the top XX differentiall-expressed genes (DEGs) between a cluster $y_1$ in the query and a cluster $y_2$ in the reference dataset.\n",
    "\n",
    "When looking at the label distance matrix, we want to check if we have a predictive signal that can help with cell-type assignments, e.g., that celltypes which one would expect to match across datasets, are also associated with a low distance (high similarity). Note, that those matches do not need to be perfect, as the there is still the distance between individual cells $distance_{gene\\,expression} (ùòÖ_1, ùòÖ_2)$. Moreover, one does not need to make assignments based on this initial sanity check, SConnect is going to do this for you, based on the underlying optimal transport model (See `OT mappings` section later in the tutorial).\n",
    "\n",
    "To give some examples for this initial sanity check, we can see that, indeed, we have a predictive signal that already distinguishes many of the cell types from each other. E.g when looking at the `NK_CD56++`, `B_Naive` or `T4_Treg` clusters from the query dataset.\n",
    "\n",
    "As a quick sanity check, this looks all good. As both datasets are immune dataset and we would expect quite a few potential matches where the string labels are the same or quite similar."
   ]
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "label_distance = dataset_map.label_distance\n",
    "# Sort label distance matrix that best matches are on the diagonal to aid visualization\n",
    "label_distance = label_distance.loc[\n",
    "    label_distance.min(axis=1).sort_values().index.tolist(),\n",
    "    label_distance.min().sort_values().index.tolist(),\n",
    "]\n",
    "\n",
    "fig = pr.pl.mapping(label_distance, zmin=0.0, zmax=1.0, colormap=\"RdYlGn_r\", sort_by_score=False)\n",
    "fig.update_coloraxes(showscale=False)\n",
    "fig"
   ],
   "id": "4499adbbb8a11137"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Show output as PNG image that it gets rendered on GitHub\n",
    "fig.write_image(\"label_distance.png\")\n",
    "display(Image(\"label_distance.png\"))"
   ],
   "id": "f86a2e7242c58606"
  },
  {
   "cell_type": "markdown",
   "id": "a85757f1-ebe7-4477-a124-55cbcd3b8689",
   "metadata": {},
   "source": [
    "Additionally, we can look at the DEGs that were used as a basis for the similarity calcualtion to see based on which genes the similarity between clusters is calcualted and whether those genes are biologically meaningful. Let's look at two examples:\n",
    "* `cM` &rarr; `CD14+_Monocyte`: Here we can see many CD14+ Monocyte markers, e.g.: `CD14`, `S100A12`, `S100A8`, `S100A9`\n",
    "*  `T4_Treg` &rarr; `Treg`: Here we can see many CD4 T Regulatory cell markers, e.g.: `FOXP3`, `RTKN2`, `IL32`, `CTLA4`, `IL2RA`\n",
    "\n",
    "Again, we can see that the method input, indeed, makes sense."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "6d0e9bdb-9a6d-4590-8418-b6e2ca5a12ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rank_query</th>\n",
       "      <th>logFC_query</th>\n",
       "      <th>rank_ref</th>\n",
       "      <th>logFC_ref</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S100A8</th>\n",
       "      <td>0</td>\n",
       "      <td>4.81</td>\n",
       "      <td>0</td>\n",
       "      <td>5.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S100A9</th>\n",
       "      <td>1</td>\n",
       "      <td>4.80</td>\n",
       "      <td>1</td>\n",
       "      <td>5.54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S100A12</th>\n",
       "      <td>2</td>\n",
       "      <td>4.55</td>\n",
       "      <td>5</td>\n",
       "      <td>4.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LYZ</th>\n",
       "      <td>3</td>\n",
       "      <td>4.19</td>\n",
       "      <td>2</td>\n",
       "      <td>4.98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCN1</th>\n",
       "      <td>4</td>\n",
       "      <td>4.15</td>\n",
       "      <td>3</td>\n",
       "      <td>4.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD14</th>\n",
       "      <td>5</td>\n",
       "      <td>4.02</td>\n",
       "      <td>7</td>\n",
       "      <td>3.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VCAN</th>\n",
       "      <td>6</td>\n",
       "      <td>3.91</td>\n",
       "      <td>4</td>\n",
       "      <td>4.32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SERPINA1</th>\n",
       "      <td>7</td>\n",
       "      <td>3.69</td>\n",
       "      <td>10</td>\n",
       "      <td>3.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CSTA</th>\n",
       "      <td>8</td>\n",
       "      <td>3.40</td>\n",
       "      <td>15</td>\n",
       "      <td>3.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MNDA</th>\n",
       "      <td>9</td>\n",
       "      <td>3.34</td>\n",
       "      <td>16</td>\n",
       "      <td>3.50</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          rank_query  logFC_query  rank_ref  logFC_ref\n",
       "gene                                                  \n",
       "S100A8             0         4.81         0       5.68\n",
       "S100A9             1         4.80         1       5.54\n",
       "S100A12            2         4.55         5       4.23\n",
       "LYZ                3         4.19         2       4.98\n",
       "FCN1               4         4.15         3       4.39\n",
       "CD14               5         4.02         7       3.86\n",
       "VCAN               6         3.91         4       4.32\n",
       "SERPINA1           7         3.69        10       3.64\n",
       "CSTA               8         3.40        15       3.50\n",
       "MNDA               9         3.34        16       3.50"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "COLUMNS_TO_SHOW = [\"rank_query\", \"logFC_query\", \"rank_ref\", \"logFC_ref\"]\n",
    "\n",
    "with pd.option_context(\"display.precision\", 2):\n",
    "    display(dataset_map.DEGs_label_distance[\"cM\"][\"CD14+_Monocyte\"][COLUMNS_TO_SHOW])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9faafe25-02d2-4626-a0cc-23e948326fd7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rank_query</th>\n",
       "      <th>logFC_query</th>\n",
       "      <th>rank_ref</th>\n",
       "      <th>logFC_ref</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>FOXP3</th>\n",
       "      <td>0</td>\n",
       "      <td>3.06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RTKN2</th>\n",
       "      <td>1</td>\n",
       "      <td>2.93</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IL32</th>\n",
       "      <td>2</td>\n",
       "      <td>2.84</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTLA4</th>\n",
       "      <td>3</td>\n",
       "      <td>2.54</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IL2RA</th>\n",
       "      <td>4</td>\n",
       "      <td>2.51</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAL</th>\n",
       "      <td>5</td>\n",
       "      <td>2.35</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD3D</th>\n",
       "      <td>6</td>\n",
       "      <td>2.26</td>\n",
       "      <td>10.0</td>\n",
       "      <td>2.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TIGIT</th>\n",
       "      <td>7</td>\n",
       "      <td>2.20</td>\n",
       "      <td>8.0</td>\n",
       "      <td>2.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNFRSF4</th>\n",
       "      <td>8</td>\n",
       "      <td>2.16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LEF1</th>\n",
       "      <td>9</td>\n",
       "      <td>2.07</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.39</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         rank_query  logFC_query  rank_ref  logFC_ref\n",
       "gene                                                 \n",
       "FOXP3             0         3.06       0.0       3.31\n",
       "RTKN2             1         2.93       3.0       2.53\n",
       "IL32              2         2.84       2.0       2.56\n",
       "CTLA4             3         2.54       1.0       2.78\n",
       "IL2RA             4         2.51       6.0       2.21\n",
       "MAL               5         2.35       7.0       2.15\n",
       "CD3D              6         2.26      10.0       2.04\n",
       "TIGIT             7         2.20       8.0       2.11\n",
       "TNFRSF4           8         2.16       NaN        NaN\n",
       "LEF1              9         2.07       4.0       2.39"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with pd.option_context(\"display.precision\", 2):\n",
    "    display(dataset_map.DEGs_label_distance[\"T4_Treg\"][\"Treg\"][COLUMNS_TO_SHOW])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1fee69cb-b80c-4d1b-9366-e2088dbb5725",
   "metadata": {},
   "source": [
    "## 4. Fit SConnect model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5d448771-a563-4640-b6a4-b9db22942ab2",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_map.init_problem(tau_a=1.0, tau_b=1.0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9111e4f-796f-4247-a6e0-b56956b2edc4",
   "metadata": {},
   "source": [
    "**Explanation of method parameters:**\n",
    "* `tau_*`: imbalance parameter of the optimal transport problem\n",
    "  * If one expects a low overlap between cell type labels of the two datasets, this can be set lower; a starting point would be to set ùúè  between 0.95 and 0.975.\n",
    "  * Note: If tau < 1., the mapping/similarity scores can be larger than 1. One should increase the unbalancedness parameter tau if values are far higher than 1.\n",
    "  * Recommended default: 1.0\n",
    "  * Recommended default (low overlap in cell type annotations): 0.95 - 0.975\n",
    "\n",
    "**See OTT-JAX documentation for more details:**\n",
    "* https://ott-jax.readthedocs.io/en/latest/problems/_autosummary/ott.problems.linear.linear_problem.LinearProblem.html#ott.problems.linear.linear_problem.LinearProblem\n",
    "* https://ott-jax.readthedocs.io/en/latest/tutorials/linear/800_Unbalanced_OT.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c002a5ac-396e-4ec9-a30d-0175d804c5f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  6%|‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñã                                                                                                                                       | 12/200 [4:37:10<72:22:23, 1385.87s/it, error: 8.019218e-04]\n"
     ]
    }
   ],
   "source": [
    "dataset_map.solve()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c069611b-104d-4e12-8dd8-4b94b64d1d5b",
   "metadata": {},
   "source": [
    "## 5. Interpreting results of SConnect"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40d61992-cc5b-450f-b3e6-c4356753a126",
   "metadata": {},
   "source": "### OT mappings"
  },
  {
   "cell_type": "markdown",
   "id": "ce80946c-e7ac-4e53-b55f-397f3f344ff5",
   "metadata": {},
   "source": [
    "This the main output of SConnect, the cluster_mapping matrix shows where the cells aggregated by cell-type cluster from the query dataset get mapped to in the reference dataset. Thus, showing to the user which clusters are most similar."
   ]
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "cluster_mapping = dataset_map.compute_mapping()",
   "id": "93498e8b8e8e1778"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "fig = pr.pl.mapping(cluster_mapping)\n",
    "# Interactive plotly plot isn't shown on GitHub\n",
    "fig"
   ],
   "id": "bf25f372a2e16bc6"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Show output as PNG image that it gets rendered on GitHub\n",
    "fig.write_image(\"cluster_mapping.png\")\n",
    "display(Image(\"cluster_mapping.png\"))"
   ],
   "id": "cc733e038aa46611"
  },
  {
   "cell_type": "markdown",
   "id": "49b92a53-d760-4c72-89fc-f564aef69a83",
   "metadata": {},
   "source": "### Cluster distance"
  },
  {
   "cell_type": "markdown",
   "id": "a5229635-dcd7-495f-be7a-8d251375c593",
   "metadata": {},
   "source": [
    "As an additional output, we can also look at the distance between cell-type clusters. This shows to the user how \"close\" or similar individual clusters are. This can be used as an additional sanity check, two clusters might be most similar based on the cluster_mapping matrix. However, they might still have a high distance between them. This should be seen as an indication to the user that such a match should be investigated in more detail."
   ]
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "cluster_distance = dataset_map.compute_distance()",
   "id": "782fa016d24b5c20"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Sort cluster distances by mapping scores as well\n",
    "cluster_distance = cluster_distance.loc[\n",
    "    cluster_mapping.max(axis=1).sort_values(ascending=False).index.tolist(),\n",
    "    cluster_mapping.max().sort_values(ascending=False).index.tolist(),\n",
    "]\n",
    "fig = pr.pl.distance(cluster_distance)\n",
    "# Interactive plotly plot isn't shown on GitHub\n",
    "fig"
   ],
   "id": "ecd0dd305b0fa5dc"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Show output as PNG image that it gets rendered on GitHub\n",
    "fig.write_image(\"cluster_distance.png\")\n",
    "display(Image(\"cluster_distance.png\"))"
   ],
   "id": "6884001ee8324d22"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ddda60c-0192-4894-9f0e-1fd75e6adb43",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
